e2e/various-suite/frontend-sandbox-datasource.spec.ts

```diff
@@ -0,0 +1,152 @@
+      beforeEach(() => {
+        cy.window().then((win) => {
+          win.localStorage.setItem('grafana.featureToggles', 'pluginsFrontendSandbox=0');
+        });
+      });
+
+      it('Should not wrap the query editor in a sandbox wrapper', () => {
+        e2e.pages.Explore.visit();
+        e2e.components.DataSourcePicker.container().should('be.visible').click();
+        cy.contains(DATASOURCE_TYPED_NAME).scrollIntoView().should('be.visible').click();
+
+        cy.wait(300); // wait to prevent false positives because cypress checks too fast
+        cy.get(`div[data-plugin-sandbox="${DATASOURCE_ID}"]`).should('not.exist');
+      });
+
+      it('Should accept values when typed', () => {
+        e2e.pages.Explore.visit();
+        e2e.components.DataSourcePicker.container().should('be.visible').click();
+        cy.contains(DATASOURCE_TYPED_NAME).scrollIntoView().should('be.visible').click();
+
+        const valueToType = 'test' + random(100);
+
+        cy.get('[data-testid="sandbox-query-editor-query-input"]').should('not.be.disabled');
+        cy.get('[data-testid="sandbox-query-editor-query-input"]').type(valueToType);
+        cy.get('[data-testid="sandbox-query-editor-query-input"]').should('have.value', valueToType);
+      });
+    });
+
+    describe('Sandbox enabled', () => {
+      beforeEach(() => {
+        cy.window().then((win) => {
+          win.localStorage.setItem('grafana.featureToggles', 'pluginsFrontendSandbox=1');
+        });
+      });
+
+      it('Should wrap the query editor in a sandbox wrapper', () => {
+        e2e.pages.Explore.visit();
+        e2e.components.DataSourcePicker.container().should('be.visible').click();
+        cy.contains(DATASOURCE_TYPED_NAME).scrollIntoView().should('be.visible').click();
+
+        cy.get(`div[data-plugin-sandbox="${DATASOURCE_ID}"]`).should('exist');
+      });
+
+      it('Should accept values when typed', () => {
+        e2e.pages.Explore.visit();
+        e2e.components.DataSourcePicker.container().should('be.visible').click();
+        cy.contains(DATASOURCE_TYPED_NAME).scrollIntoView().should('be.visible').click();
+
+        const valueToType = 'test' + random(100);
+
+        cy.get('[data-testid="sandbox-query-editor-query-input"]').should('not.be.disabled');
+        cy.get('[data-testid="sandbox-query-editor-query-input"]').type(valueToType);
+        cy.get('[data-testid="sandbox-query-editor-query-input"]').should('have.value', valueToType);
+
+        // typing the query editor should reflect in the url
+        cy.url().should('include', valueToType);
+      });
+    });
+  });
+
+  afterEach(() => {
+    e2e.flows.revertAllChanges();
+  });
+
+  after(() => {
+    // delete the datasource
+    e2e.flows.login(Cypress.env('USERNAME'), Cypress.env('PASSWORD'), true);
+    cy.request({
+      url: `${Cypress.env('BASE_URL')}/api/datasources/name/${DATASOURCE_TYPED_NAME}`,
+      method: 'DELETE',
+      failOnStatusCode: false,
+    });
```

Comment:
Sorry one more thing. If this is a prerequisite for the tests above to work, you may want to move it to the `before` hook as there's no guarantee this code is ever executed.

Reply:
This is a cleanup method. it deletes the created datasource. that's why is on the after.

Reply:
Yes. But if you for example refresh cypress in the middle of the test, the after hook is never executed and the next time the tests run they (may) not work because your app was in a bad state. That's why you want to put all prerequisites in the `before` hook or in `beforeEach`.

Reply:
In the middle of a test that didn't run the after there are two scenarios:

* In CI: This test will fail and the pipeline will fail, clean up is irrelevant
* in local: creating the new datasource won't work because one with the same name already exists.

Keep in mind this cleanup is more a nice to have than a requirement for this test to work. I didn't feel like leaving a datasource created around for the rest of the tests. this part of the code is not necessary at all for any of these tests to work.